net_sink.go

```diff
@@ -305,34 +318,36 @@ func (s *netSink) run() {
 			n := len(s.outc)
 			for i := 0; i < n && s.conn != nil; i++ {
 				buf := <-s.outc
-				s.writeToConn(buf)
+				if err := s.writeToConn(buf); err != nil {
+					s.retryc <- buf
+					continue
+				}
 				putBuffer(buf)
 			}
 			close(done)
 		case buf := <-s.outc:
-			s.writeToConn(buf)
+			if err := s.writeToConn(buf); err != nil {
+				s.retryc <- buf
+				continue
```

Comment:
Need to return the buffer to the pool, I guess you have two options here: add another call to `putBuffer` before `continue` or remove `continue` (next iteration of the loop won't do anything because `conn` will be `nil`).

Reply:
I think in this case we really do want to continue without `putBuffer`: the buffer is still in use (we just sent it to `retryc`) and retryc will call `putBuffer` when it's done. If we call putBuffer here, we risk some other writer messing with the buffer before the `retryc` processing completes.

Reply:
ohhhhhhh that's a good point, I totally missed that the buffer is the thing sent in the channel